RealmSwift/Projection.swift

```diff
@@ -0,0 +1,1342 @@
+            }
+            return .change(newProjection, projectedPropertyChanges)
+        case .deleted:
+            return .deleted
+        }
+    }
+}
+
+/// ``Projection`` is a light weight model of the original Realm ``Object`` or ``EmbeddedObject``.
+/// You can use `Projection` as a view model to minimize boilerplate.
+///
+/// Example of usage:
+/// ```swift
+/// public class Person: Object {
+///     @Persisted var firstName = ""
+///     @Persisted var lastName = ""
+///     @Persisted var address: Address? = nil
+///     @Persisted var friends = List<Person>()
+///     @Persisted var reviews = List<String>()
+/// }
+///
+/// public class Address: EmbeddedObject {
+///     @Persisted var city: String = ""
+///     @Persisted var country = ""
+/// }
+///
+/// class PersonProjection: Projection<Person> {
+///     @Projected(\Person.firstName) var firstName
+///     @Projected(\Person.lastName.localizedUppercase) var lastNameCaps
+///     @Projected(\Person.address.city) var homeCity
+///     @Projected(\Person.friends.projectTo.firstName) var friendsFirstName: ProjectedList<String>
+/// }
+/// ```
+///  ### Supported property types
+/// Projection can transform the original `@Persisted` properties in several ways:
+/// - `Passthrough` - `Projection`'s property will have same name and type as original object. See `PersonProjection.firstName`.
+/// - `Rename` - Projection's property will have same type as original object just with the new name.
+/// - `Keypath resolution` - you can access the certain properties of the projected `Object`. See `PersonProjection.lastNameCaps` and `PersonProjection.homeCity`.
+/// - `Collection mapping` - `List` and `MutableSet`of `Object`s or `EmbeddedObject`s  can be projected as a collection of primitive values.
+///     See `PersonProjection.friendsFirstName`.
+/// - `Exclusion` - all properties of the original Realm object that were not defined in the projection model will be excluded from projection.
+///     Any changes happened on those properties will not trigger a change notification for the `Projection`.
+///     You still can access the original `Object` or `EmbeddedObject` and observe notifications directly on it.
+/// - note: each `@Persisted` property can be `@Projected` in different ways in the same Projection class.
+/// Each `Object` or `EmbeddedObject` can have sevaral projections of same or different classes at once.
+///
+/// ### Querying
+/// You can retrieve all Projections of a given type from a Realm by calling the `objects(_:)` of Realm or `init(projecting:)`
+/// of Projection's class:
+/// ```swift
+/// let projections = realm.object(PersonProjection.self)
+/// let personObject = realm.create(Person.self)
+/// let singleProjection = PersonProjection(projecting: personObject)
+/// ```
+open class Projection<Root: ObjectBase>: RealmCollectionValue, ProjectionObservable {
+    /// The object being projected
+    public var rootObject: Root
```

Comment:
This should be `let` right?

Reply:
var, but with private setter to access `observed.rootObject[keyPath: *WritableKeyPath*]`